flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java

```diff
@@ -270,10 +270,12 @@ private void updateLastSeenWatermark(long timestamp) {
 	}
 
 	private void updateNFA(NFA<IN> nfa) throws IOException {
-		if (nfa.isEmpty()) {
-			nfaOperatorState.clear();
-		} else {
-			nfaOperatorState.update(nfa);
+		if (nfa.isNFAChanged()) {
+			if (nfa.isEmpty()) {
+				nfaOperatorState.clear();
+			} else {
+				nfaOperatorState.update(nfa);
```

Comment:
Currently for each input, we will deserialize the NFA and the `nfaChanged` field is `false` by default. So I think we don't need to do that. Thoughts?